Source/WebKit/webpushd/WebClipCache.mm

```diff
@@ -0,0 +1,253 @@
+            NSArray *array = (NSArray *)entry;
+            for (id val in array) {
+                if (![val isKindOfClass:[NSString class]] || ![val length]) {
+                    RELEASE_LOG_ERROR(Push, "WebClipCache::load failed to deserialize %{public}@: value isn't a string", path);
+                    return nil;
+                }
+            }
+        }
+
+        return entries;
+    }
+}
+
+void WebClipCache::load()
+{
+    RetainPtr entries = loadWebClipCachePropertyList(m_path);
+    if (!entries)
+        return;
+
+    for (NSArray *entry in entries.get()) {
+        NSString *bundleIdentifier = entry[0];
+        NSString *securityOriginString = entry[1];
+        NSString *webClipIdentifier = entry[2];
+
+        RetainPtr securityOriginURL = adoptNS([[NSURL alloc] initWithString:securityOriginString]);
+        auto origin = WebCore::SecurityOriginData::fromURL(URL { securityOriginURL.get() });
+        if (origin.isNull())
+            continue;
+
+        m_preferredWebClipIdentifiers.add(std::make_tuple(String { bundleIdentifier }, origin), String { webClipIdentifier });
+    }
+}
+
+void WebClipCache::persist()
+{
+    @autoreleasepool {
+        RetainPtr<NSMutableArray> entries = adoptNS([[NSMutableArray alloc] init]);
+        for (const auto& [bundleIdentifierAndOrigin, webClipIdentifier] : m_preferredWebClipIdentifiers) {
+            NSString *bundleIdentifier = (NSString *)std::get<0>(bundleIdentifierAndOrigin);
+            NSURL *url = (NSURL *)std::get<1>(bundleIdentifierAndOrigin).toURL();
+            NSString *securityOriginString = [url absoluteString];
+            if (!securityOriginString)
+                continue;
+            [entries addObject:@[bundleIdentifier, securityOriginString, (NSString *)webClipIdentifier]];
+        }
+
+        NSError *error = nil;
+        NSData *data = [NSPropertyListSerialization dataWithPropertyList:entries.get() format:NSPropertyListBinaryFormat_v1_0 options:0 error:&error];
+        if (!data) {
+            RELEASE_LOG_ERROR(Push, "WebClipCache::persist failed to serialize plist: %{public}@", error);
+            return;
+        }
+
+        if (![data writeToFile:(NSString *)m_path options:NSDataWritingAtomic error:&error]) {
+            RELEASE_LOG_ERROR(Push, "WebClipCache::persist failed to write to disk: %{public}@", error);
+            return;
+        }
+    }
+}
+
+HashSet<String> WebClipCache::visibleWebClipIdentifiers(const String& bundleIdentifier)
+{
+    HashSet<String> result;
+
+    @autoreleasepool {
+        NSArray *webClips = [UIWebClip webClips];
+
+        for (UIWebClip *webClip in webClips) {
+            NSString *identifier = webClip.identifier;
+            if (!identifier)
+                continue;
+
+            if (bundleIdentifier == "com.apple.SafariViewService"_s || ![webClip respondsToSelector:@selector(trustedClientBundleIdentifiers)] || [webClip.trustedClientBundleIdentifiers containsObject:(NSString *)bundleIdentifier]) {
```

Comment:
Should the identifier be added if the webClip does not respond to `trustedClientBundleIdentifiers`?

Reply:
Yes, it's for staging only until enough builds have that API available. Can add a comment.